src/Symfony/Component/HttpClient/UriTemplateHttpClient.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Component\HttpClient\UriTemplate\CallbackUriTemplateExpander;
+use Symfony\Component\HttpClient\UriTemplate\UriTemplateExpanderInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+use Symfony\Contracts\HttpClient\ResponseStreamInterface;
+
+class UriTemplateHttpClient implements HttpClientInterface
+{
+    private HttpClientInterface $client;
+    private UriTemplateExpanderInterface $expander;
+
+    private array $variables = [];
+
+    public function __construct(
+        HttpClientInterface $client = null,
+        UriTemplateExpanderInterface $expander = null,
+    ) {
+        $this->client = $client ?? HttpClient::create();
+        $this->expander = $expander ?? CallbackUriTemplateExpander::createFromPopularVendors();
+    }
+
+    public function request(string $method, string $url, array $options = []): ResponseInterface
+    {
+        if ($this->variables && null !== ($options['base_uri'] ?? null)) {
+            $options['base_uri'] = $this->expander->expand($options['base_uri'], $this->variables);
+        }
+
+        return $this->client->request(
+            $method,
+            $this->variables ? $this->expander->expand($url, $this->variables) : $url,
+            $options,
+        );
+    }
+
+    public function stream(iterable|ResponseInterface $responses, float $timeout = null): ResponseStreamInterface
+    {
+        return $this->client->stream($responses, $timeout);
+    }
+
+    public function withOptions(array $options): static
+    {
+        if ($this->variables && null !== ($options['base_uri'] ?? null)) {
+            $options['base_uri'] = $this->expander->expand($options['base_uri'], $this->variables);
+        }
+
+        $clone = clone $this;
+        $clone->client = $this->client->withOptions($options);
+
+        return $clone;
+    }
+
+    public function withVariables(array $variables): static
```

Comment:
what about using options to deal with variables instead?

`$client->request('GET', $url, ['vars' => [...])`

Reply:
I studied this option but it means we would add a new option just for one type of client, so we would add it to the `extra` key I guess. It feels verbose and less discoverable :man_shrugging:

Reply:
And that would make it work when combined with other decorators (while `withVariables` would work only when it is the outer client)